src/Compilers/CSharp/Portable/CSharpParseOptions.cs

```diff
@@ -178,26 +180,30 @@ public new CSharpParseOptions WithFeatures(IEnumerable<KeyValuePair<string, stri
                 throw new ArgumentNullException(nameof(features));
             }
 
-            // there are currently no parse options for experimental features
-            if (System.Linq.Enumerable.Any(features))
-            {
-                throw new ArgumentException("Experimental features are not supported", nameof(features));
-            }
+            var newFeatures = features.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value);
 
-            return this;
+            return new CSharpParseOptions(this) { _features = newFeatures };
         }
 
         public override IReadOnlyDictionary<string, string> Features
         {
             get
             {
-                // there are currently no parse options for experimental features
-                return new Dictionary<string, string>();
+                if (_features == null)
+                {
+                    _features = new Dictionary<string, string>();
```

Comment:
@agocke I changed it to what @jaredpar mentioned, putting ImmutableDictionary.Empty in the constructor. I didn't know about Empty, and did it this lazy way to avoid an allocation if it was just going to be overwritten in WithFeatures anyway. Avoids the interlocked problem that way.